README.md


# RIVM Toolbox for Microbiome data analyses  
In the last years at the RIVM much research has been done on data management and on the development of statistical methods for “data wrangling” (such as cleaning, transforming, merging data…), as well as for visualization of complex multivariable associations. The toolbox will implement these already available developments in combination with the new requirements in a user-friendly format, using the R language, and making them available to the scientific community within the RIVM as well as (inter)nationally.

## biomeViz  
`biomeViz` focus is towards data visualization for microbiota. Data visualization is done mostly using the ggplot2 and related family of R packages.  
The package is <span style="color:red">undergoing development</span>.  

## Related tools  

* [biomeStats](https://github.com/RIVM-IIV-Microbiome/biomeStats)    
* [biomeUtils](https://rivm-iiv-microbiome.github.io/biomeUtils/articles/Introduction.html)    





NEWS.md
# biomeViz 0.0.06

* Added plotScatterViz.

# biomeViz 0.0.05

* Changes to theme_biomeViz.

# biomeViz 0.0.04

* Added a `themes` to the package.


R/utils.R
#' Standard ggplot2 theme for RIVM Toolbox
#'
#' @name Themes
#'
#' @details A standard theme for ggplot2 to produce similar aesthetics for
#'          RIVM Toolbox.
#'
#' @param base_size Base size
#' @param base_family Base family
#'
#' @importFrom ggplot2 element_text element_rect element_blank theme_bw theme
#'   rel %+replace%
#' @importFrom grid unit
#'
#' @examples
#' library(biomeUtils)
#' library(biomeViz)
#' library(microbiome)
#' library(ggplot2)
#' library(dplyr)
#'
#' ps <- FuentesIliGutData %>%
#'   microbiome::aggregate_taxa("Genus") %>%
#'   microbiome::transform("compositional")
#'
#' plotTaxaRaincloud(ps,
#'                   taxa ="Bacteroides",
#'                   group_samples_by = "ILI",
#'                   opacity = 0.25,
#'                   shape_point = 21) + # combine with ggplot2 for improvements
#'   labs(y = "Relative abundance",
#'        x = "",
#'        subtitle = "Raincloud plot") +
#'   theme_biomViz_minimal()+
#'   scale_fill_biomeViz(palette = "viz3") +
#'   scale_color_biomeViz(palette = "viz3")
#'

NULL

#' @rdname Themes
#' @aliases theme_biomViz
#' @importFrom ggplot2 element_text element_rect element_blank theme_bw theme margin
#' @export
theme_biomViz <- function(base_size = 12, base_family = "") {
  half_line <- base_size/2
  theme_bw(base_size = base_size, base_family = base_family) %+replace%
    theme(
      text = element_text(colour = "black",
                          face = "plain",
                          size = base_size,
                          vjust = 0.5,
                          hjust = 0.5,
                          lineheight = 1),
      panel.grid.minor = element_blank(),
      #panel.border = element_rect(fill = NA, colour = "#181818"),
      axis.text = element_text(colour="#303030", size = rel(1)),
      axis.title = element_text(colour = "grey10", size = rel(1.2)),
      axis.ticks = element_line(color = "grey50"),
      axis.ticks.length = unit(.3, "lines"),
      strip.background = element_blank(),
      strip.text = element_text(size=rel(1.0)),
      # strip.switch.pad.grid = unit(half_line/2, "pt"),
      # strip.switch.pad.wrap = unit(half_line/2, "pt"),
      #strip.placement = "outside",
      legend.title = element_text(colour = "#303030", size = rel(1.2), face = "bold", hjust=0),
      legend.key.size = unit(1.2, "lines"),
      legend.text = element_text(size = rel(1.2), colour = "#303030"),
      legend.key = element_rect(colour = NA, fill = NA),
      legend.background = element_rect(colour = NA, fill = NA),
      plot.title = element_text(size = rel(1.0), color = "grey10", face = "bold",
                                vjust = 1, margin = margin(b = half_line)),
      plot.subtitle = element_text(size = rel(1.0), color = "grey10", hjust = 0,
                                   vjust = 1, margin = margin(b = half_line)),
      plot.caption = element_text(size = rel(0.8), margin = margin(t = 15))
    )
}


#' @rdname Themes
#' @aliases theme_biomViz_minimal
#' @importFrom ggplot2 element_text element_rect element_blank theme_minimal theme
#'   rel %+replace% element_line margin
#' @export
theme_biomViz_minimal <- function(base_size = 11, base_family = "") {
  half_line <- base_size/2
  theme_minimal(base_size = base_size, base_family = base_family) %+replace%
    theme(
      panel.grid.major = element_line(colour="#f0f0f0"),
      panel.grid.minor = element_blank(),
      axis.ticks.length = unit(half_line / 2.2, "pt"),
      axis.ticks = element_line(colour = "#303030")  ,
      strip.background = element_rect(fill = NA, colour = NA),
      strip.text.x = element_text(colour = "#303030", size = rel(1.2), face = "bold"),
      strip.text.y = element_text(colour = "#303030", size = rel(1.2), face = "bold"),
      axis.text = element_text(colour="#303030", size = rel(1)),
      axis.title = element_text(colour = "grey10", size = rel(1.2)),
      axis.line = element_line(colour="#303030"),
      legend.title = element_text(colour = "#303030", size = rel(1.2), face = "bold", hjust=0),
      panel.border = element_blank(),
      legend.key.size = unit(1.2, "lines"),
      legend.text = element_text(size = rel(1.2), colour = "#303030"),
      legend.key = element_rect(colour = NA, fill = NA),
      legend.background = element_rect(colour = NA, fill = NA),
      plot.title = element_text(colour = "#303030", size = rel(1.2),hjust = 0, margin=margin(0,0,10,0)),
      plot.subtitle = element_text(colour = "#303030", size = rel(1),hjust = 0, margin=margin(0,0,10,0)),
      plot.margin=unit(c(10,5,5,5),"mm")
    )
}

#' @rdname Themes
#' @aliases theme_biomViz_bw
#' @importFrom ggplot2 element_text element_rect element_blank theme_minimal theme
#'   rel %+replace% element_line margin
#' @export
theme_biomViz_bw <- function(base_size = 11, base_family = "") {
  half_line <- base_size/2
  theme_bw(base_size = base_size, base_family = base_family) %+replace%
    theme(panel.background = element_rect(fill = "#F8F8F8", color = "#303030"),
      panel.grid.major = element_line(colour= "#DCDCDC"),
      panel.grid.minor = element_blank(),
      axis.ticks.length = unit(half_line / 2.2, "pt"),
      axis.ticks = element_line(colour = "#303030")  ,
      strip.background = element_rect(fill = "#F8F8F8", colour = "#303030"),
      strip.text.x = element_text(colour = "#303030", size = rel(1), face = "bold"),
      strip.text.y = element_text(colour = "#303030", size = rel(1), face = "bold", angle = -90),
      axis.text = element_text(colour="#303030", size = rel(1)),
      axis.title = element_text(colour = "grey10", size = rel(1.2)),
      #axis.line = element_line(colour="#303030"),
      legend.title = element_text(colour = "#303030", size = rel(1.2), face = "bold", hjust=0),
      panel.border = element_blank(),
      legend.key.size = unit(1.2, "lines"),
      legend.text = element_text(size = rel(1.2), colour = "#303030"),
      legend.key = element_rect(colour = NA, fill = NA),
      legend.background = element_rect(colour = NA, fill = NA),
      plot.title = element_text(colour = "#303030", size = rel(1.2),hjust = 0, vjust = 1,margin=margin(0,0,10,0)),
      plot.subtitle = element_text(colour = "#303030", size = rel(1),hjust = 0, vjust = 1, margin=margin(0,0,10,0)),
      plot.margin=unit(c(10,5,5,5),"mm")
    )
}



# Themes setup from https://drsimonj.svbtle.com/creating-corporate-colour-palettes-for-ggplot2
# https://twitter.com/drsimonj
#' @rdname Themes
#' @aliases scale_fill_biomeViz_summer
#' @importFrom scales manual_pal
#' @importFrom ggplot2 discrete_scale
#' @export
scale_fill_biomeViz_summer <- function(...){
  discrete_scale("fill","biomeViz_summer",manual_pal(values = c("#E56997","#BD97CB",
                                                                "#66D2D6","#F5631A","#FFA303",
                                                                "#444440","#07BB9C","#4120A9",
                                                                "#FEC8A7","#116530","#D7D4DD")), ...)

}

#' @rdname Themes
#' @aliases scale_colour_biomeViz_summer
#' @importFrom scales manual_pal
#' @importFrom ggplot2 discrete_scale
#' @export
scale_colour_biomeViz_summer <- function(...){
  discrete_scale("colour","biomeViz_summer",manual_pal(values = c("#E56997","#BD97CB",
                                                                  "#66D2D6","#F5631A","#FFA303",
                                                                  "#444440","#07BB9C","#4120A9",
                                                                  "#FEC8A7", "#116530","#D7D4DD")), ...)

}

#' Function to extract biomeViz colors as hex codes
#'
#' @param ... Character names of biomeViz_colors
#' @rdname Themes
#' @aliases biomeViz_cols
#' @export
biomeViz_cols <- function(...) {
  cols <- c(...)

  if (is.null(cols))
    return (biomeViz_colors)

  biomeViz_colors[cols]
}

#' List of biomeViz colors
#' @rdname Themes
#' @export
biomeViz_colors <- c(
  `darkgreen` = "#3d6721",
  `beigeyellow` = "#e9c46a",
  `beigeorange` = "#f4a261",
  `beigebrown` = "#936639",
  `beigetblue` = "#457b9d",
  `beigegreen` = "#b5e48c",
  `beigegrey` = "#adb5bd")

#biomeViz_cols("beigegreen")


#' biomeViz palettes
#' @rdname Themes
#' @export
biomeViz_palettes <- list(
  `summer`  = biomeViz_cols("beigegreen","beigebrown","beigeyellow","beigeorange","beigetblue","beigegrey"),
  `summer3`  = biomeViz_cols("beigetblue","beigeyellow","beigeorange"),
  `viz3` = biomeViz_cols("darkgreen","beigebrown","beigetblue")
)

#' Return function to interpolate a biomeViz color palette
#'
#' @param palette Character name of palette in biomeViz_palettes
#' @param reverse Boolean indicating whether the palette should be reversed
#' @param ... Additional arguments to pass to colorRampPalette()
#' @importFrom grDevices colorRampPalette
#' @rdname Themes
#' @export
biomeViz_pal <- function(palette = "summer", reverse = FALSE, ...) {
  pal <- biomeViz_palettes[[palette]]

  if (reverse) pal <- rev(pal)

  colorRampPalette(pal, ...)
}

#biomeViz_pal("summer")
#' Color scale constructor for biomeViz colors
#'
#' @param palette Character name of palette in biomeViz_palettes
#' @param discrete Boolean indicating whether color aesthetic is discrete or not
#' @param reverse Boolean indicating whether the palette should be reversed
#' @param ... Additional arguments passed to discrete_scale() or
#'            scale_color_gradientn(), used respectively when discrete is TRUE or FALSE
#'
#' @importFrom ggplot2 discrete_scale scale_color_gradientn
#' @rdname Themes
#' @aliases scale_color_biomeViz
#' @export
scale_color_biomeViz <- function(palette = "summer", discrete = TRUE, reverse = FALSE, ...) {
  pal <- biomeViz_pal(palette = palette, reverse = reverse)

  if (discrete) {
    discrete_scale("colour", paste0("biomeViz_", palette), palette = pal, ...)
  } else {
    scale_color_gradientn(colours = pal(256), ...)
  }
}

#' Fill scale constructor for biomeViz colors
#'
#' @param palette Character name of palette in biomeViz_palettes
#' @param discrete Boolean indicating whether color aesthetic is discrete or not
#' @param reverse Boolean indicating whether the palette should be reversed
#' @param ... Additional arguments passed to discrete_scale() or
#'            scale_fill_gradientn(), used respectively when discrete is TRUE or FALSE
#'
#' @importFrom ggplot2 discrete_scale scale_fill_gradientn
#' @rdname Themes
#' @aliases scale_fill_biomeViz
#' @export
scale_fill_biomeViz <- function(palette = "summer", discrete = TRUE, reverse = FALSE, ...) {
  pal <- biomeViz_pal(palette = palette, reverse = reverse)

  if (discrete) {
    discrete_scale("fill", paste0("biomeViz_", palette), palette = pal, ...)
  } else {
    scale_fill_gradientn(colours = pal(256), ...)
  }
}


R/plotByGroup.R
#' Visualize one categorical column and one numeric column
#'
#' @name plotByGroup
#'
#' @details A simple plotting utility for visualization of one categorical
#'          and numerical variable. Provided a categorical variable and numerical
#'          variable present in \code{sample_data} of a
#'          \code{\link{phyloseq-class}} object, a comparative visualization
#'          can be done.
#'
#' @param x \code{\link{phyloseq-class}} object.
#'
#' @param x.factor A column in \code{sample_data} to compare that is a factor
#'
#' @param y.numeric A column in \code{sample_data} to compare that is a numeric
#'
#' @return a \code{ggplot} object.
#'
#' @examples
#' library(biomeUtils)
#' library(biomeViz)
#' library(dplyr)
#' library(microbiome)
#' library(ggplot2)
#' ps <- FuentesIliGutData
#' # calculate Shannon diversity using microbiome::diversity and add it to
#' # the sample_data in phyloseq using biomeUtils::mutateSampleData()
#' ps <- mutateSampleData(ps,
#'                        Shannon = microbiome::diversity(ps, "shannon")[,1])
#' plotByGroup(ps,
#'             x.factor="ILI",
#'             y.numeric = "Shannon") +
#'          geom_point(aes(color=ILI),
#'                     position = position_jitter(width = 0.1)) +
#'          scale_color_manual(values=c("steelblue", "brown3", "grey70"))
#'
#' @author Sudarshan A. Shetty
#'
#' @references
#' Shetty SA (2021). Data visualization for microbiome analytics.
#' \url{https://github.com/microsud/biomeViz}
#'
#' @importFrom dplyr %>%
#' @importFrom biomeUtils getSampleTibble
#' @importFrom rlang sym
#'
#' @export

plotByGroup <- function(x,
                        x.factor = NULL,
                        y.numeric = NULL){

  if(is.null(x.factor) | is.null(y.numeric)) {
    stop("Please specify x.factor and y.numeric")
  }

  num <- NULL
  data.plot <- x %>%
    getSampleTibble()

  vars.use <- c(x.factor,y.numeric)
  if( !vars.use[1] %in% colnames(data.plot) | !vars.use[2] %in% colnames(data.plot)) {
    stop("variables not found in sample_data")
  }

  # check
  .check_factor_numeric(data.plot,
                        x.factor = x.factor,
                        y.numeric = y.numeric)

  nsamples.data <- .get_nsamples_by_group(x, x.factor)

  data.plot <- data.plot %>%
    dplyr::left_join(nsamples.data, by = x.factor) %>%
    dplyr::mutate(myaxis = paste0(!!sym(x.factor), "\n", "n=", num))



  p <- data.plot %>%
    ggplot2::ggplot(ggplot2::aes_string(x="myaxis",
                                        y=y.numeric)) +
    ggplot2::geom_boxplot(
      ggplot2::aes_string(fill=x.factor),
      width = .1,
      outlier.shape = NA,
      alpha=0.5) +
    xlab("") +
    theme_minimal()
  return(p)
}


# Helper
#' @importFrom rlang sym
#' @importFrom dplyr %>% pull
.check_factor_numeric <- function(data.plot, x.factor, y.numeric){

  if(is.null(x.factor) | is.null(y.numeric)) {
    stop("Please specify x.factor and y.numeric")
  }

  cat.vals  <- data.plot %>%
    pull(!!sym(x.factor))
  num.vals  <- data.plot %>%
    pull(!!sym(y.numeric))
  if(!is.factor(cat.vals) | !is.numeric(num.vals)) {
    stop("x.factor must be a factor and y.numeric must be numeric values")
  }

}

#' @importFrom biomeUtils getSampleTibble
.get_nsamples_by_group <- function(x, x.factor) {

  if(!x.factor %in% phyloseq::sample_variables(x)) {
    stop("x.factor variable is not present in",
         "sample_data")
  }

  n <- num <- NULL
  nsamples.data <- x %>%
    getSampleTibble() %>%
    dplyr::group_by(!!sym(x.factor)) %>%
    dplyr::summarize(num=n())
  return(nsamples.data)
}



R/plotScatterViz.R
#' Scater Plot
#'
#' @name plotScatterViz
#'
#' @details A simple plotting utility for visualization of two numerical
#'          variables. Provided two numerical
#'          variable present in \code{sample_data} of a
#'          \code{\link{phyloseq-class}} object, a scatter plot is returned.
#'
#' @param x \code{\link{phyloseq-class}} object.
#'
#' @param x_numeric A column in \code{sample_data} to compare that is a numeric
#'
#' @param y_numeric A column in \code{sample_data} to compare that is a numeric
#'
#' @param color_var A column in \code{sample_data} to color points
#'
#' @param point_size \code{geom_point}. Default is 3
#'
#' @param point_opacity \code{ggplot2} alpha value. Default is 0.5
#'
#' @param point_shape \code{ggplot2} point shape value. Default is 21
#'
#' @param side_panel Logical. If side density to plot. Default is TRUE
#'
#' @param side_panel_scale \code{ggside} ggside.panel.scale requires a scalar numeric.
#'                         default is 0.3
#'
#' @return a \code{ggplot} object.
#'
#' @examples
#' library(biomeUtils)
#' library(biomeViz)
#' library(dplyr)
#' library(microbiome)
#' library(ggplot2)
#' ps <- FuentesIliGutData
#' plotScatterViz(ps, x_numeric = "age", y_numeric = "BMI",
#'                color_var = "ILI") +
#'                scale_fill_manual(values=c("steelblue", "brown3", "grey70"))+
#'                scale_fill_manual(values=c("steelblue", "brown3", "grey70"))
#' @author Sudarshan A. Shetty
#' @references
#' Shetty SA (2021). Data visualization for microbiome analytics.
#' \url{https://github.com/microsud/biomeViz}
#'
#' @export
NULL

plotScatterViz <- function(x,
                           x_numeric = NULL,
                           y_numeric = NULL,
                           color_var = NULL,
                           point_size = 3,
                           point_opacity = 0.5,
                           point_shape = 21,
                           side_panel=TRUE,
                           side_panel_scale = 0.3){

  p <- getSampleTibble(x) |>
    ggplot2::ggplot(ggplot2::aes_string(x_numeric, y_numeric,fill=color_var))

  p <- p +
    ggplot2::geom_point(size=point_size,
                        alpha=point_opacity,
                        shape=point_shape) +
    theme_biomViz()

  if(side_panel){
    p <- p +
      ggside::geom_xsidedensity(ggplot2::aes_string(fill=color_var,color= color_var),
                                alpha = point_opacity) +
      ggside::geom_ysidedensity(ggplot2::aes_string(fill=color_var,color= color_var),
                                alpha = point_opacity)+
      theme(ggside.panel.scale = side_panel_scale) +
      ggside::scale_ysidex_continuous(guide = ggplot2::guide_axis(angle = 90), minor_breaks = NULL)
  }
  return(p)

}


R/plotTidyHeatmap.R
#' Creates a Heatmap using `TidyHeatmap` and `ComplexHeatmap`
#'
#' @name plotTidyHeatmap
#'
#' @details A TidyHeatmap plotting wrapper for \code{\link{phyloseq-class}} object.
#'
#' @param x \code{\link{phyloseq-class}} object.
#'
#' @param select_taxa Features to plot.
#'
#' @param group_samples_by A column in \code{sample_data} to group tiles.
#'
#' @param add_taxa_label Logical. Default is TRUE.
#'
#' @param ... Arguments forwarded to tidyHeatmap::heatmap and ComplexHeatmap::Heatmap.
#'
#' @return a \code{heatmap} object.
#'
#' @examples
#' library(biomeUtils)
#' library(biomeViz)
#' library(microbiome)
#' library(dplyr)
#'
#' ps <- FuentesIliGutData %>%
#'          microbiome::transform("compositional")
#' select_taxa <- findTopTaxa(ps, top= 10, method="mean")
#'
#' plotTidyHeatmap(ps, select_taxa = select_taxa,
#'                 group_samples_by = "ILI",
#'                 add_taxa_label = FALSE,
#'                 cluster_rows = FALSE,
#'                 .scale = "none",
#'                 transform = NULL,
#'                 palette_value = c("red", "white", "blue"))
#'
#' @author Sudarshan A. Shetty
#'
#' @references
#' Shetty SA (2021). Data visualization for microbiome analytics.
#' \url{https://github.com/microsud/biomeViz}
#'
#' @importFrom dplyr %>% case_when across everything
#' @importFrom tidyr replace_na
#' @importFrom biomeUtils getSampleTibble getTaxaTibble getAbundanceTibble
#' @importFrom rlang sym
#' @importFrom stringr str_c
#' @importFrom phyloseq sample_variables
#'
#' @export
plotTidyHeatmap <- function(x,
                            select_taxa = NULL,
                            group_samples_by = NULL,
                            add_taxa_label = TRUE,
                            ...){

  FeatureID <- Features <- SampleID <- n <- num <- NULL
  sample_variables <- value <- NULL

  if(is.null(select_taxa) | is.null(group_samples_by)) {
    stop("Please specify select_rows and or group_samples_by")
  }
  if ( !is(x, "phyloseq") ){
    stop("input must be an phyloseq object.")
  }

  # Make first data
  plot.data <- x %>%
    getAbundanceTibble(select_rows = select_taxa) %>%
    tidyr::pivot_longer(-FeatureID, names_to = "SampleID")

  if(add_taxa_label){

    tax_df <- getTaxaTibble(x) %>%
      #mutate(Species = split_species(Species, n = num_species)) %>%
      dplyr::mutate(Features =
                      case_when(
                        is.na(Class)  ~ str_c(FeatureID, ":p_", Phylum),
                        is.na(Order)  ~ str_c(FeatureID, ":c_", Class),
                        is.na(Family)  ~ str_c(FeatureID, ":o_", Order),
                        is.na(Genus)   ~ str_c(FeatureID, ":f_", Family),
                        is.na(Species) ~ str_c(FeatureID, ":g_", Genus),
                        TRUE ~ str_c(FeatureID, ":" , Genus, " ", Species)
                      )
      ) %>%
      dplyr::filter(FeatureID %in% select_taxa) %>%
      dplyr::mutate(across(everything(), ~replace_na(.x, "Unclassified")))

    plot.data <- plot.data %>%
      dplyr::left_join(tax_df, by = "FeatureID")

  } else {

    tax_df <- getTaxaTibble(x, select_rows = select_taxa)  %>%
      dplyr::mutate(across(everything(), ~replace_na(.x, "Unclassified")))

    plot.data <- plot.data %>%
      dplyr::left_join(tax_df,
                       by = "FeatureID") %>%
      dplyr::rename(Features = "FeatureID")

  }

  if( !group_samples_by %in% sample_variables(x) ) {
    stop("group_samples_by value not found in sample_data")
  }

  plot.data <- plot.data %>%
    dplyr::left_join(getSampleTibble(x), by = "SampleID") %>%
    dplyr::group_by(!!sym(group_samples_by))


  p <- plot.data %>%
    tidyHeatmap::heatmap(
      .column = SampleID,
      .row = Features,
      .value = value,
      ...)

  return(p)
}


R/plotPCoA.R
#' PCoA plot
#'
#' @name plotPCoA
#'
#' @details A Principal Coordinates Analysis for \code{\link{phyloseq-class}} object.
#'          To visualize similarities or dissimilarities between samples in 2D ordination.
#'          This function extends the \code{phyloseq} ordination plots to include
#'          taxa that correlate with choosen axis and plots them along with a
#'          side boxplot for comparing inter-sample variation within groups.
#'
#' @param x \code{\link{phyloseq-class}} object.
#'
#' @param group_var A column in \code{sample_data} to compare. This is also the
#'                  variable used to colour.
#'
#' @param ord_method Ordination method, currently tested PCoA.
#'
#' @param dist_method Distance method, currently tested Bray-Curtis.
#'
#' @param seed Random seed number \code{set.seed}.
#'
#' @param cor_method Correlation method. Default is \code{Spearman}.
#'
#' @param padj_cutoff Cut-off for multiple testing. Default is 0.05.
#'
#' @param padj_method Method for multiple testing. Default is fdr.
#'
#' @param arrows Logical. If arrows for taxa with correlation to axis are to be
#'               plotted. Default is TRUE.
#'
#' @param label_col Color of labels. Default is grey30.
#'
#' @param plot_centroids Logical. To plot centroids or not. Default is TRUE.
#'
#' @param add_side_box Logical. To plot side boxplots or not. Default is TRUE.
#'
#' @param axis_plot Which axis to plot. Default is first two.
#'
#' @param point_shape Shape of the points. Default is 21.
#'
#' @param point_alpha Opacity of points. Default is 0.5
#'
#' @param verbose Logical. Messages to print. Default is TRUE.
#'
#' @param ... Additional arguments to pass to vegan's adonis function.
#'
#' @return a \code{ggplot2} object.
#'
#' @examples
#' library(biomeUtils)
#' library(dplyr)
#' library(ggside)
#' ps <- FuentesIliGutData %>%
#'   microbiome::transform("compositional") %>%
#'   mutateTaxaTable(FeatureID = taxa_names(FuentesIliGutData))
#' plotPCoA(x =ps,
#'          group_var = "ILI",
#'          ord_method = "PCoA",
#'          dist_method = "bray",
#'          seed = 1253,
#'          cor_method = "spearman",
#'          verbose = TRUE,
#'          padj_cutoff = 0.05,
#'          padj_method = "fdr",
#'          arrows = TRUE,
#'          label_col = "grey30",
#'          plot_centroids = TRUE,
#'          add_side_box = TRUE,
#'          axis_plot = c(1:2),
#'          point_shape = 21,  # point_shape
#'          point_alpha = 0.5) +
#'   scale_color_manual(values = c("#3d6721", "#a86826", "#006c89")) +
#'   scale_fill_manual(values = c("#3d6721", "#a86826", "#006c89"))
#'
#' @author Sudarshan A. Shetty
#' @references
#' Shetty SA (2021). Data visualization for microbiome analytics.
#' \url{https://github.com/microsud/biomeViz}
#'
#' @export
NULL
plotPCoA <- function(x,
                     group_var,
                     ord_method,
                     dist_method,
                     seed,
                     cor_method = "spearman",
                     padj_cutoff = 0.05,
                     padj_method = "fdr",
                     arrows = TRUE,
                     label_col = "grey30",
                     plot_centroids = TRUE,
                     add_side_box = TRUE,
                     axis_plot = c(1:2),
                     point_shape = 21,  # point_shape
                     point_alpha = 0.5,
                     verbose = TRUE,
                     ...){

  id.type <- FeatureID <- Axis.1 <- Axis.2 <- NULL
  ord.dat <- .get_ord_info(x,
                           group_var,
                           ord_method,
                           dist_method,
                           seed,
                           verbose= verbose,
                           ...)

  ordplot.dat.all <- phyloseq::plot_ordination(x,
                                               ord.dat$ordDF,
                                               justDF = TRUE,
                                               type = "split")

  # get samples ordination
  ordplot.dat.samples <- subset(ordplot.dat.all, id.type =="Samples")

  # get sample centroids
  axis.a <- colnames(ordplot.dat.samples)[1]
  axis.b <- colnames(ordplot.dat.samples)[1]

  centroids.samples <- ordplot.dat.samples %>%
    dplyr::group_by(!!rlang::sym(group_var)) %>%
    dplyr::summarise(centroid.axis.1 = mean(!!rlang::sym(axis.a)),
                     centroid.axis.2 = mean(!!rlang::sym(axis.b)))

  # Get axis 1 and 2 variation
  evals1 <- round(ord.dat$ordDF$values$Eigenvalues[1] / sum(ord.dat$ordDF$values$Eigenvalues) * 100, 2)
  evals2 <- round(ord.dat$ordDF$values$Eigenvalues[2] / sum(ord.dat$ordDF$values$Eigenvalues) * 100, 2)

  # Get Taxa PC correlation
  tax.pc.cor.dat <- .pc_taxa_correlation(x,
                                         ordplot.dat.samples= ordplot.dat.samples,
                                         cor_method,
                                         axis_plot,
                                         padj_cutoff,
                                         padj_method)


  ordplot.dat.taxa <- subset(ordplot.dat.all, id.type == "Taxa")

  #head(ordplot.dat.taxa)
  names(tax.pc.cor.dat) <- NULL
  axis.a.tax <- as.data.frame(tax.pc.cor.dat[1])$rowname
  axis.b.tax <- as.data.frame(tax.pc.cor.dat[2])$rowname

  .check_taxa_axis(axis.a.tax,axis.b.tax)

  select_tax <- c(axis.a.tax,axis.b.tax)

  ordplot.dat.taxa <- ordplot.dat.taxa %>%
    dplyr::filter(FeatureID %in% select_tax)

  axis_a_names <- colnames(ordplot.dat.samples)[1]
  axis_b_names <- colnames(ordplot.dat.samples)[2]
  # start plot
  p <- ggplot2::ggplot(ordplot.dat.samples,
                       ggplot2::aes_string(x = axis_a_names, y = axis_b_names)) +
    ggplot2::geom_point(ggplot2::aes_string(color = group_var,
                                            fill = group_var),
                        shape= point_shape,  # point_shape
                        alpha= point_alpha)  # point_alpha

  if(add_side_box){

    p  <- p +
      ggside::geom_xsideboxplot(ggplot2::aes_string(y = axis_a_names,
                                                    color= group_var,
                                                    fill = group_var),
                                shape = point_shape,  # point_shape
                                alpha = point_alpha,
                                orientation = "y") +
      ggside::geom_ysideboxplot(ggplot2::aes_string(x = axis_a_names,
                                                    color = group_var,
                                                    fill = group_var),
                                shape= point_shape,  # point_shape
                                alpha = point_alpha,
                                orientation = "x") +
      ggside::scale_ysidex_continuous(guide = ggplot2::guide_axis(angle = 90),
                                      minor_breaks = NULL) +
      ggside::scale_xsidey_discrete() +
      ggside::scale_ysidex_discrete()

  }

  if(arrows){
    p <- p +
      ggplot2::geom_segment(data=ordplot.dat.taxa,
                            ggplot2::aes(x=0, xend=Axis.1*4,
                                         y=0, yend=Axis.2*4),
                            colour = label_col,              # arrow_color
                            size=0.5,                        # arrow_size
                            arrow = ggplot2::arrow(length= ggplot2::unit(1, "mm"),
                                                   type="closed")) +
      ggrepel::geom_label_repel(data=ordplot.dat.taxa,
                                ggplot2::aes(x=Axis.1*4, y=Axis.2*4, # scale_vars
                                             label=FeatureID), # taxa_label
                                colour=label_col,     # label_color
                                #label.padding=.1,
                                size=3,               # label_size
                                max.overlaps = 100000,
                                alpha = 0.75,         # label_opacity
                                box.padding = 0.80,
                                point.padding = 0.5,
                                na.rm=TRUE)
  }

  p + ggplot2::labs(x = paste0("PCoA (", evals1, "%)"),
                    y = paste0("PCoA (", evals2, "%)"))
}

# Get ordination

.get_ord_info <- function(x,
                          group_var = NULL,
                          ord_method = "PCoA",
                          dist_method ="bray",
                          seed = 123,
                          verbose = verbose,
                          ...){

  if(is.null(group_var) || !any(sample_variables(x) %in% group_var)){
    stop("Provide a valid grouping variable")
  }

  if (!is(x, "phyloseq")){
    stop("Ïnput x must be a phyloseq object")
  }

  ord <- phyloseq::ordinate(x, method = ord_method, distance = dist_method)

  dist.mat <- phyloseq::distance(x, method = dist_method)

  group.levs <- microbiome::meta(x)[,group_var]

  if(is.na(seed) || is.null(seed)){
    stop("Provide a random number, see `set.seed`")
  }
  set.seed(seed)
  if (verbose){
    message(paste0("Random number for permutation analysis ...\n", seed))
  }

  permanova.res <- vegan::adonis(dist.mat~ group.levs, ...)

  return(list(ordDF = ord, permanovaRes = permanova.res))
}


# Taxa PC correlation

.pc_taxa_correlation <- function(x,
                                 ordplot.dat.samples= ordplot.dat.samples,
                                 cor_method = cor_method,
                                 axis_plot = axis_plot,
                                 padj_cutoff = padj_cutoff,
                                 padj_method = padj_method){

  # select axis
  pcs <- ordplot.dat.samples[,axis_plot]
  pc.cor.taxa <- NULL
  for (i in colnames(pcs)){

    suppressWarnings(
      correlate.pcs <- apply(microbiome::abundances(x),1,function(x)
      stats::cor.test(x, pcs[,i], method = cor_method))
    )

    pc.corr <- sapply(correlate.pcs,"[[",4)
    pc.pval <- sapply(correlate.pcs,"[[",3)
    pc.padj <- stats::p.adjust(pc.pval, method = padj_method)

    #otu.cor.pc <- abundances(x)[which(pc.padj < alpha & abs(pc.corr) >= quantile(abs(pc.corr),0.95,na.rm = T)),]
    sig.corr <- pc.corr[which(pc.padj < padj_cutoff & abs(pc.corr) >= stats::quantile(abs(pc.corr),0.95,na.rm = T))]
    sig.p <- pc.padj[which(pc.padj < padj_cutoff & abs(pc.corr) >= stats::quantile(abs(pc.corr),0.95,na.rm = T))]

    taxa.sig.axis.1 <- tibble::tibble(rowname=names(sig.corr),
                                      corr = sig.corr,
                                      p.adj = sig.p)

    taxa.sig.axis.1$rowname <- gsub(".rho$", "",taxa.sig.axis.1$rowname)
    pc.cor.taxa[[i]] <- taxa.sig.axis.1

  }
  return(pc.cor.taxa)
}


.check_taxa_axis <- function(axis.a.tax,axis.b.tax){
  if (length(axis.a.tax) == 0 && length(axis.b.tax) == 0){
    warning("Both of the choosen axis in `axis_plot` have no taxa satisfying criteria to plot")
  }

  if (length(axis.a.tax) == 0){
    warning("First of the choosen axis in `axis_plot` has no taxa satisfying criteria to plot")
  }

  if (length(axis.b.tax) == 0){
    warning("Second of the choosen axis in `axis_plot` has no taxa satisfying criteria to plot")
  }
}
